Move code from src-d/ml

Guillemdb opened the pull request:
This is the first stage of the refactor of `src-d/ml` into `ml-core`. 

- The code is moved to ml-core and the imports are fixed.
- The tests pass.
- Most of the CI pipeline works correctly.
- Added black for styling (@vmarkovtsev approved)

zurk commented:
Well, as soon as it is almost the same files we have in src-d/ml can you be more specific how do you change it if you did. What I should pay more attention to?

Guillemdb commented:
The files are the same, the only stuff I did was:
- Move al UAST related files into the uast folder.
- Change the imports to match the new project name.
- Remove references to pyspark in the README files.
- Change the references in the CI/setup files from srcd.ml to ml_core.
- Apply black as a code formatter.

The functionality has not been changes, I just changed the file locations.

zurk commented on Dockerfile:
```diff
@@ -0,0 +1,31 @@
+
+COPY . ml_core/
+RUN cd ml_core && pip3 install -e .
+# ENTRYPOINT ["analyzer"]
```
We do not have ENTRYPOINT now, let's remove it and add later if we have one.
```suggestion
```

zurk commented on Makefile:
```diff
@@ -0,0 +1,33 @@
+		srcd/ml_core -m unittest discover
+
+.PHONY: bblfsh-start
+bblfsh-start:
```
let's not include for now it because bblfsh versions always depend on the project.

zurk commented on README.md:
```diff
@@ -1,4 +1,4 @@
-# MLonCode research playground [![PyPI](https://img.shields.io/pypi/v/sourced-ml.svg)](https://pypi.python.org/pypi/sourced-ml) [![Build Status](https://travis-ci.org/src-d/ml.svg)](https://travis-ci.org/src-d/ml) [![Docker Build Status](https://img.shields.io/docker/build/srcd/ml.svg)](https://hub.docker.com/r/srcd/ml) [![codecov](https://codecov.io/github/src-d/ml/coverage.svg)](https://codecov ...
+# MLonCode research playground [![PyPI](https://img.shields.io/pypi/v/sourced-ml.svg)](https://pypi.python.org/pypi/sourced-ml) [![Build Status](https://travis-ci.org/src-d/ml.svg)](https://travis-ci.org/src-d/ml) [![Docker Build Status](https://img.shields.io/docker/build/srcd/ml.svg)](https://hub.docker.com/r/srcd/ml) [![codecov](https://codecov.io/github/src-d/ml/coverage.svg)](https://codecov ...
```
This readme is not about ml-core. For now, I suggest to create a new tiny README.md and make a proper one later. Let's keep Contribution, License, Glossary, Docker image. Algorithms have wrong links; Installation is different, so let's remove for now. 

I also suggest using `MLonCode Core Library` as a title and add a small relevant description. For example: 

```
Library for machine learning on the source code. Provides commonly used algorithms and tools to process the code-related data such as Babelfish's UASTs, plain code text, etc.
```
Feel free to change it as you like, it is an example.

zurk commented on ml_core/__init__.py:
```diff
@@ -0,0 +1,12 @@
+"""MLonCode research playground."""
```
```suggestion
"""MLonCode Core Library."""
```

zurk commented on ml_core/__init__.py:
```diff
@@ -0,0 +1,12 @@
+except ImportError:
+    pass
+
+__version__ = "0.8.3"
```
```suggestion
__version__ = "0.0.1"
```

zurk commented on ml_core/__init__.py:
```diff
@@ -0,0 +1,12 @@
+"""MLonCode research playground."""
+import pkg_resources
+
+pkg_resources.declare_namespace(__name__)
```
We should do it only in case we share the same namespace.
I suggest doing so (see review summary message), but in case we will not, we should delete this line

Guillemdb replied:
Is it ok if I leave this there for now till Vadim answers, or should I remove it and add it later if needed?

zurk replied:
yes, let's wait for Vadim's input.

zurk commented on ml_core/algorithms/id_embedding.py:
```diff
@@ -0,0 +1,42 @@
+import numpy
+
+# todo(guillemdb): maybe refactor in mining
```
I think that everything that does not depend on `gitbase` project should be in this repo. So let's keep it.
```suggestion
```

zurk commented on ml_core/extractors/identifiers.py:
```diff
@@ -0,0 +1,20 @@
+# TODO (Guillemdb): fix imports
```
Looks like they are fixed, aren't they?
Here are everywhere

Guillemdb replied:
Let me fix all of this in the next PR. I already corrected most of the minor style improvements I could find in a different branch. Once I fix everything which is not code related I will push it.

zurk replied:
can you explain what is wrong with these imports?

Guillemdb replied:
Nothing, I forgot to remove the TODO.

zurk commented on setup.py:
```diff
@@ -0,0 +1,71 @@
+    download_url="https://github.com/src-d/ml-core",
+    packages=find_packages(exclude=("ml_core.tests",)),
+    namespace_packages=["ml_core"],
+    entry_points={"console_scripts": ["srcml=ml_core.__main__:main"]},
```
We do not have `ml_core.__main__` let's delete it.
```suggestion
```

zurk commented on setup.py:
```diff
@@ -0,0 +1,71 @@
+    extras_require={
+        "tf": ["tensorflow>=1.0,<2.0"],
+        "tf_gpu": ["tensorflow-gpu>=1.0,<2.0"],
+        "pandas": ["pandas>=0.24.1,<1.0"],
```
there are no pandas usage in the code. Let's drop it.
```suggestion
```

zurk commented on setup.py:
```diff
@@ -0,0 +1,71 @@
+    tests_require=["docker>=3.6.0,<4.0"],
+    package_data={
+        "": ["LICENSE.md", "README.md"],
+        "sourced": ["ml_core/transformers/languages.yml"],
```
no transformers. 
```suggestion
```

zurk commented on setup.py:
```diff
@@ -0,0 +1,71 @@
+        "": ["LICENSE.md", "README.md"],
+        "sourced": ["ml_core/transformers/languages.yml"],
+    },
+    python_requires=">=3.4",
```
I think we should drop Python 3.4 as we did with other projects.
```suggestion
    python_requires=">=3.5",
```

Guillemdb commented:
I pushed a new commit with the requested changes and the .travis.yml file. I corrected the flake8 issues in another branch, if you allow me to ignore them in the next PR they will be fixed.

Guillemdb commented:
The build is failing because I remove the references to bblfsh on the MAKEFILE. Should I add it again so the build passes?

zurk commented:
oh, sure. In this case, we should keep it.

Guillemdb commented:
When I run the test on my local repository I find that only one of them is failing: 

```python 

======================================================================
FAIL: test_result (ml_core.tests.test_uast_to_id_distance.Uast2IdLineDistanceTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/guillem/srcd/ml-core/ml_core/tests/test_uast_to_id_distance.py", line 141, in test_result
    self.assertEqual(res, correct)
AssertionError: Lists differ: [(('__spec__', 'ModuleSpec'), 0), (('__spec__',[1679 chars], 1)] != [(('__package__', 'ModuleSpec'), 2), (('__spec_[2142 chars], 1)]

First differing element 0:
(('__spec__', 'ModuleSpec'), 0)
(('__package__', 'ModuleSpec'), 2)

Second list contains 14 additional elements.
First extra element 55:
(('utmain', 'collections'), 2)

+ [(('__package__', 'ModuleSpec'), 2),
- [(('__spec__', 'ModuleSpec'), 0),
? ^

+  (('__spec__', 'ModuleSpec'), 0),
? ^

   (('__spec__', 'ModuleSpec'), 1),
   (('__spec__', 'ModuleSpec'), 1),
+  (('__spec__', 'ModuleSpec'), 2),
   (('__spec__', '__package__'), 0),
+  (('collections', 'ModuleSpec'), 1),
   (('collections', 'ModuleSpec'), 2),
   (('collections', '__package__'), 1),
   (('collections', '__spec__'), 1),
+  (('collections', '__spec__'), 2),
   (('modules', '__package__'), 1),
   (('modules', '__spec__'), 1),
   (('modules', 'collections'), 2),
   (('namedtuple', 'ModuleSpec'), 0),
   (('namedtuple', 'ModuleSpec'), 1),
+  (('namedtuple', 'ModuleSpec'), 1),
   (('namedtuple', 'ModuleSpec'), 2),
   (('namedtuple', 'ModuleSpec'), 2),
   (('namedtuple', '__package__'), 1),
+  (('namedtuple', '__package__'), 2),
   (('namedtuple', '__spec__'), 1),
   (('namedtuple', '__spec__'), 1),
+  (('namedtuple', '__spec__'), 2),
+  (('namedtuple', '__spec__'), 2),
   (('namedtuple', 'collections'), 0),
-  (('namedtuple', 'collections'), 2),
?                                  ^

+  (('namedtuple', 'collections'), 1),
?                                  ^

   (('namedtuple', 'modules'), 2),
   (('setup_logging', 'modelforge.logs'), 0),
   (('setup_logging', 'setup'), 1),
   (('sys', '__package__'), 1),
   (('sys', '__spec__'), 1),
   (('sys', 'collections'), 2),
   (('sys', 'modelforge.logs'), 2),
   (('sys', 'modules'), 0),
   (('sys', 'namedtuple'), 2),
   (('sys', 'setup_logging'), 2),
   (('utmain', 'ModuleSpec'), 0),
   (('utmain', 'ModuleSpec'), 1),
   (('utmain', 'ModuleSpec'), 1),
   (('utmain', 'ModuleSpec'), 1),
   (('utmain', 'ModuleSpec'), 2),
+  (('utmain', 'ModuleSpec'), 2),
+  (('utmain', 'ModuleSpec'), 2),
   (('utmain', '__package__'), 0),
   (('utmain', '__package__'), 0),
   (('utmain', '__package__'), 1),
   (('utmain', '__spec__'), 0),
   (('utmain', '__spec__'), 0),
   (('utmain', '__spec__'), 0),
   (('utmain', '__spec__'), 1),
   (('utmain', '__spec__'), 2),
   (('utmain', 'collections'), 1),
   (('utmain', 'collections'), 1),
   (('utmain', 'collections'), 2),
+  (('utmain', 'collections'), 2),
   (('utmain', 'modules'), 0),
   (('utmain', 'modules'), 1),
   (('utmain', 'modules'), 1),
   (('utmain', 'namedtuple'), 1),
   (('utmain', 'namedtuple'), 1),
   (('utmain', 'namedtuple'), 1),
   (('utmain', 'namedtuple'), 2),
+  (('utmain', 'namedtuple'), 2),
+  (('utmain', 'namedtuple'), 2),
+  (('utmain', 'namedtuple'), 2),
   (('utmain', 'sys'), 0),
   (('utmain', 'sys'), 1),
   (('utmain', 'sys'), 1)]

```
Does the file `test/source/example.py` depend on the project from where it is called?
Does this test somehow fail for something I did wrong?

However, during the CI, a lot of tests fail with a similar error to this one>

```python
ERROR: test_greatest2 (ml_core.tests.test_df.DocumentFrequenciesTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/ml_core/ml_core/tests/test_df.py", line 10, in setUp
    self.model = DocumentFrequencies().load(source=paths.DOCFREQ)
  File "/usr/local/lib/python3.6/dist-packages/modelforge/model.py", line 93, in load
    cache_dir = os.path.join(self.cache_dir(), self.NAME)
  File "/usr/local/lib/python3.6/dist-packages/modelforge/model.py", line 338, in cache_dir
    raise RuntimeError("modelforge is not configured; look at modelforge.configuration. "
RuntimeError: modelforge is not configured; look at modelforge.configuration. Depending on your objective you may or may not want to create a modelforgecfg.py file which sets VENDOR and the rest.
```
The `modelforgecfg.py` is created, but somehow the path to the files is `"/ml_core/ml_core/tests/test_df.py"`, and the folder `ml_core` is duplicated. Do you have any idea of why this happens? 

I am creating an additional copy of `ml_core` and moving the whole project inside that folder, but I'm not sure where. Could it be in the [L29 of the Dockerfile](https://github.com/src-d/ml-core/blob/382ff195316fec1fea2f840dfacc6627c579fdb9/Dockerfile#L29)?

zurk commented:
I will check tomorrow.

zurk commented:
I add my changes to the commit. Now everything works fine. 
There are several things that @Guillemdb before I can approve it I ask you to do several more things:
1. Namespace thing is approved by Vadim (I asked him in DM). 
    1. Put everything you have in `core_ml` to `sourced/ml/core`. Address all `core_ml` entries in code accordingly.
    1. Since we do not have the second package yet (`sourced.ml.mining`) keep `__init__.py` empty from `pkg_resources.declare_namespace(__name__)` we will add it later. If we add it now, it is not possible to test it. 
1. Search for all `sourced.ml` entries in the PR and address them accordingly. 
2. Add `flake8-commas==2.0.0` to requrements-lint.txt and address new style issues.

zurk commented:
things that should be addressed in the next PRs (@Guillemdb you can convert this list to issues after we merge this PR):
1. Add a directory for documentation and documentation-related checks to Travis as we have in https://github.com/src-d/style-analyzer.
2. Add the deployment stage to Travis
3. If we use black styler we should have a check in Travis that it was applied.
4. Add dependabot as we have in https://github.com/src-d/ml (see its PRs: https://github.com/src-d/ml/pull/406)

Guillemdb commented:
I finally managed to fix the CI and upload the following changes:

- The namespace of the library from `ml_core` to `sourced.ml.core`.
- Updated CI and configuration files accordingly to comply with the new name.
- I didn't implement `flake8-commas==2.0.0` because [black already takes care of it](https://test-black.readthedocs.io/en/style-guide/style_guide/trailing_commas.html). (I would like to add black as a pre-commit hook in future PRs.)
- Renamed the folder `sourced.ml.core.tests.models` to `sourced.ml.core.tests.models_tensorflow` to avoit conflicts with the file `sourced.ml.core.tests.models.py`.
- Skipped the test `test_serialize` from `tests.models_tensorflow.test_tensorflow` due to an unknown error with stack-trace:
```python
======================================================================
ERROR: test_serialize (ml.core.tests.models_tensorflow.test_tensorflow.TensorFlowModelTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/virtualenv/python3.5.6/lib/python3.5/site-packages/sourced/ml/core/tests/models_tensorflow/test_tensorflow.py", line 18, in test_serialize
    TensorFlowModel().construct(graphdef=gd).save(buffer)
  File "/home/travis/virtualenv/python3.5.6/lib/python3.5/site-packages/modelforge/model.py", line 390, in save
    raise ValueError("series must be specified")
ValueError: series must be specified
```

zurk commented:
> I didn't implement flake8-commas==2.0.0 because black already takes care of it. (I would like to add black as a pre-commit hook in future PRs.)

Ok. One thing, you should add not a pre-commit hook but another style check (see https://github.com/src-d/ml-core/pull/1#issuecomment-484558774). I am fine if you do it in the next PR. In the best case, black check should completely replace flake8.

zurk commented on README.md:
```diff
@@ -1,95 +1,7 @@
-# MLonCode research playground [![PyPI](https://img.shields.io/pypi/v/sourced-ml.svg)](https://pypi.python.org/pypi/sourced-ml) [![Build Status](https://travis-ci.org/src-d/ml.svg)](https://travis-ci.org/src-d/ml) [![Docker Build Status](https://img.shields.io/docker/build/srcd/ml.svg)](https://hub.docker.com/r/srcd/ml) [![codecov](https://codecov.io/github/src-d/ml/coverage.svg)](https://codecov ...
+# MLonCode Cora Library
```
```suggestion
# MLonCode Core Library
```

Guillemdb replied:
Done

smola commented:
@Guillemdb @vmarkovtsev It would be desirable to preserve history or, at least, separate commits for code import and code changes, so it's clear the point at which it's imported and what changes were made.

Guillemdb commented:
So far there is no changes to the code functionality besides fixing the Tensorflow test, I was waiting for this PR to be merged to make any change to the core-ml functionality if needed.

smola commented:
@Guillemdb But there were non-functional changes, right?

Guillemdb commented:
Yeah, all the changed were non-functional, what I did was:
- Resetting the .md files to a minified version that only mentions the `core-ml`project.
- Copy-pasting the files to the new repo.
- Fixing the imports to match the new namespace speciffications.
- Apply black formatting to minimize future diffs.
- Change the CI so it's defined as in the `lookout-sdk-ml` project (as @vmarkovtsev suggested).

vmarkovtsev commented:
@Guillemdb It is impossible for me to review 118 files. I will comment and the thing will turn into an unmanageable mess. Can you please split all the changes into feature-full 5-6 PRs and submit them one by one. It is fine that each following PR depends on the previous ones.

1. Move the text files
2. Add CI. The repo is empty so it passes. CI should additionally test packaging btw.
3. Add the old code. Pass CI.
4. Add black formatting. Yes, that is important, (3) must be **exactly** the same old code apart from the CI fixes.
5. Add the rest of the files, e.g. autodoc generation
